pggan/models/modules.py

```diff
@@ -5,78 +5,111 @@
+            x = module(x)
+
+            if block_len == 1:
+                x = self.toRGBs[f'from_{self._get_out_channels(x.size(-1))}'](x)
+                break
+
+            if block_idx == block_len - 2 and self.fade_in_weight < 1:
+                residual_x = nn.Upsample(scale_factor=2, mode='nearest')(x)
+                residual_x = self.toRGBs[f'from_{self._get_out_channels(residual_x.size(-1) >> 1)}'](residual_x)
+                residual_x = residual_x.mul(1-self.fade_in_weight)
+
+            if block_idx == block_len - 1:
+                x = self.toRGBs[f'from_{self._get_out_channels(x.size(-1))}'](x)
+                if self.fade_in_weight < 1:
+                    x = x.mul(self.fade_in_weight).add(residual_x)
+
+        return x
+
+    def _get_toRGB(self, in_channels):
+        return nn.Sequential(
+            nn.Conv2d(
+                in_channels=in_channels,
+                out_channels=3,
+                kernel_size=1,
+                padding=0,
+            ),
         )
 
-    def forward(
+    def _get_out_channels(self, resolution):
+        return min(self.maximum_channel, 2**(15-(len(bin(resolution))-2)))
+
+    def set_output_image_size(self, image_size):
+        if image_size > self.current_output_image_size:
+            in_channels = min(self.maximum_channel,
+                              self._get_out_channels(image_size) << 1)
+            out_channels = self._get_out_channels(image_size)
+
+            # 특정 중간 레이어에서 rgb 아웃풋낼 수 있도록 모든 out_channels 사이즈에 맞춰서 toRGB 레이어 생성
+            if f'from_{out_channels}' not in self.toRGBs.keys():
+                self.toRGBs[f'from_{out_channels}'] = self._get_toRGB(in_channels=out_channels)
+
+            new_conv_block = nn.Sequential(OrderedDict([
+                ('upsample', nn.Upsample(scale_factor=2, mode='nearest')),
+                ('conv_1', nn.Conv2d(in_channels, out_channels, 3, 1, 1)),
+                ('lrelu_1', nn.LeakyReLU(self.leakiness)),
+                ('pnorm_1', self.pixel_norm),
+                ('conv_2', nn.Conv2d(out_channels, out_channels, 3, 1, 1)),
+                ('lrelu_2', nn.LeakyReLU(self.leakiness)),
+                ('pnorm_2', self.pixel_norm),
+            ]))
+
+            self.conv_blocks.append(new_conv_block)
+            self.current_output_image_size = image_size
+        else:
+            pass
+
+
+class PixelwiseNorm(nn.Module):
+    def __init__(
         self,
-        x
-    ) -> torch.Tensor:
-        x = self.head(x)
-        x = self.body(x)
-        x = self.tail(x)
-        return x
+        method: str = 'L2'
+    ) -> None:
+        super(PixelwiseNorm, self).__init__()
+        self.method = method
+
+    def forward(self, x) -> torch.Tensor:
+        if self.method == 'L2':
+            order = 2
+        elif self.method == 'LC':
+            order = x.size(1)
+        else:
+            raise ValueError
+
+        return x.div(x.norm(order, dim=1, keepdim=True) + 1e-8)
```

Comment:
x.norm() 후에 epsilon을 더해준 것은 본 논문에서 소개한 방법과 다르다고 생각합니다.

예를 들어 order가 2인 경우에, 논문에서는:
-  (x,y)의 pixel에 대해서 모든 feature map의 값의 제곱의 평균을 구하고
-  epsilon을 더한 후
- sqrt를 계산합니다.

이 코드는,
 - (x,y)의 pixel에 대해서 모든 feature map의 값의 제곱의 평균을 구하고 (여기까지는 논문과 같음)
 - sqrt를 계산한 후,
 - epsilon으로 나눕니다.

```suggestion
        return x.div(x.pow(2).mean(dim=1, keepdim=True).add(1e-8).sqrt())
```